// typia/lib/programmers/internal/application_native.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.application_native = void 0;
var JSON_SCHEMA_PREFIX_1 = require("./JSON_SCHEMA_PREFIX");
/**
 * @internal
 */
var application_native = function (options) {
    return function (components) {
        return function (name) {
            return function (nullable) {
                var _a, _b, _c;
                var _d;
                var key = options.purpose === "ajv"
                    ? name
                    : "".concat(name).concat(nullable ? ".Nullable" : "");
                if (((_a = components.schemas) === null || _a === void 0 ? void 0 : _a[key]) === undefined) {
                    (_b = components.schemas) !== null && _b !== void 0 ? _b : (components.schemas = {});
                    (_c = (_d = components.schemas)[key]) !== null && _c !== void 0 ? _c : (_d[key] = {
                        type: "object",
                        $id: options.purpose === "ajv"
                            ? "".concat(JSON_SCHEMA_PREFIX_1.JSON_COMPONENTS_PREFIX, "/objects/").concat(key)
                            : undefined,
                        properties: {},
                        nullable: options.purpose === "swagger" ? nullable : undefined,
                    });
                }
                return {
                    $ref: "".concat(JSON_SCHEMA_PREFIX_1.JSON_COMPONENTS_PREFIX, "/objects/").concat(key),
                };
            };
        };
    };
};
exports.application_native = application_native;
