// typia/lib/utils/RandomGenerator/RandomGenerator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.datetime = exports.date = exports.pattern = exports.ipv6 = exports.ipv4 = exports.url = exports.email = exports.uuid = exports.length = exports.pick = exports.array = exports.string = exports.number = exports.bigint = exports.integer = exports.boolean = void 0;
var randexp_1 = __importDefault(require("randexp"));
var ALPHABETS = "abcdefghijklmnopqrstuvwxyz";
/* -----------------------------------------------------------
      REGULAR
  ----------------------------------------------------------- */
var boolean = function () { return Math.random() < 0.5; };
exports.boolean = boolean;
var integer = function (min, max) {
    min !== null && min !== void 0 ? min : (min = 0);
    max !== null && max !== void 0 ? max : (max = 100);
    return Math.floor(Math.random() * (max - min + 1)) + min;
};
exports.integer = integer;
var bigint = function (min, max) {
    min !== null && min !== void 0 ? min : (min = BigInt(0));
    max !== null && max !== void 0 ? max : (max = BigInt(100));
    return BigInt((0, exports.integer)(Number(min), Number(max)));
};
exports.bigint = bigint;
var number = function (min, max) {
    min !== null && min !== void 0 ? min : (min = 0);
    max !== null && max !== void 0 ? max : (max = 100);
    return Math.random() * (max - min) + min;
};
exports.number = number;
var string = function (length) {
    return new Array(length !== null && length !== void 0 ? length : (0, exports.integer)(5, 10))
        .fill(0)
        .map(function () { return ALPHABETS[(0, exports.integer)(0, ALPHABETS.length - 1)]; })
        .join("");
};
exports.string = string;
var array = function (closure, count) {
    return new Array(count !== null && count !== void 0 ? count : (0, exports.length)()).fill(0).map(function (_e, index) { return closure(index); });
};
exports.array = array;
var pick = function (array) { return array[(0, exports.integer)(0, array.length - 1)]; };
exports.pick = pick;
var length = function () { return (0, exports.integer)(0, 3); };
exports.length = length;
/* -----------------------------------------------------------
      SECIAL FORMATS
  ----------------------------------------------------------- */
var uuid = function () {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
        var r = (Math.random() * 16) | 0;
        var v = c === "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
};
exports.uuid = uuid;
var email = function () { return "".concat((0, exports.string)(10), "@").concat((0, exports.string)(10), ".").concat((0, exports.string)(3)); };
exports.email = email;
var url = function () { return "https://".concat((0, exports.string)(10), ".").concat((0, exports.string)(3)); };
exports.url = url;
var ipv4 = function () { return (0, exports.array)(function () { return (0, exports.integer)(0, 255); }, 4).join("."); };
exports.ipv4 = ipv4;
var ipv6 = function () {
    return (0, exports.array)(function () { return (0, exports.integer)(0, 65535).toString(16); }, 8).join(":");
};
exports.ipv6 = ipv6;
var pattern = function (regex) { return new randexp_1.default(regex).gen(); };
exports.pattern = pattern;
var date = function (min, max) {
    min !== null && min !== void 0 ? min : (min = 0);
    max !== null && max !== void 0 ? max : (max = Date.now() * 2);
    return new Date((0, exports.number)(min, max)).toISOString().substring(0, 10);
};
exports.date = date;
var datetime = function (min, max) {
    min !== null && min !== void 0 ? min : (min = Date.now() - 30 * DAY);
    max !== null && max !== void 0 ? max : (max = Date.now() + 7 * DAY);
    return new Date((0, exports.number)(min, max)).toISOString();
};
exports.datetime = datetime;
var DAY = 86400000;
