// typia/lib/factories/MetadataFactory.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataFactory = void 0;
var typescript_1 = __importDefault(require("typescript"));
var Metadata_1 = require("../schemas/metadata/Metadata");
var MetadataConstant_1 = require("../schemas/metadata/MetadataConstant");
var explore_metadata_1 = require("./internal/metadata/explore_metadata");
var iterate_metadata_collection_1 = require("./internal/metadata/iterate_metadata_collection");
var iterate_metadata_sort_1 = require("./internal/metadata/iterate_metadata_sort");
var ExpressionFactory_1 = require("./ExpressionFactory");
var MetadataFactory;
(function (MetadataFactory) {
    MetadataFactory.analyze = function (checker, context) {
        return function (options) {
            return function (collection) {
                return function (type) {
                    var errors = [];
                    var meta = (0, explore_metadata_1.explore_metadata)(checker)(options)(collection)(errors)(type, {
                        top: true,
                        object: null,
                        property: null,
                        nested: null,
                        escaped: false,
                        aliased: false,
                    });
                    (0, iterate_metadata_collection_1.iterate_metadata_collection)(errors)(collection);
                    (0, iterate_metadata_sort_1.iterate_metadata_sort)(collection)(meta);
                    if (options.validate)
                        errors.push.apply(errors, __spreadArray([], __read(validate(context)(options)(options.validate)(meta)), false));
                    return errors.length
                        ? {
                            success: false,
                            errors: errors,
                        }
                        : {
                            success: true,
                            data: meta,
                        };
                };
            };
        };
    };
    /**
     * @internal
     */
    MetadataFactory.soleLiteral = function (value) {
        var meta = Metadata_1.Metadata.initialize();
        meta.constants.push(MetadataConstant_1.MetadataConstant.from({
            values: [value],
            type: "string",
        }));
        return meta;
    };
    var validate = function (context) {
        return function (options) {
            return function (functor) {
                return function (meta) {
                    var visitor = {
                        functor: functor,
                        errors: [],
                        objects: new Set(),
                        arrays: new Set(),
                        tuples: new Set(),
                        aliases: new Set(),
                    };
                    validateMeta(context)(options)(visitor)(meta, {
                        object: null,
                        property: null,
                        nested: null,
                        top: true,
                        aliased: false,
                        escaped: false,
                    });
                    return visitor.errors;
                };
            };
        };
    };
    var validateMeta = function (context) {
        return function (options) {
            return function (visitor) {
                return function (meta, explore) {
                    var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f, e_7, _g, e_8, _h, e_9, _j;
                    var result = [];
                    if (context !== undefined)
                        try {
                            for (var _k = __values(meta.atomics), _l = _k.next(); !_l.done; _l = _k.next()) {
                                var atomic = _l.value;
                                try {
                                    for (var _m = (e_2 = void 0, __values(atomic.tags)), _o = _m.next(); !_o.done; _o = _m.next()) {
                                        var row = _o.value;
                                        try {
                                            for (var _p = (e_3 = void 0, __values(row.filter(function (t) { return t.validate !== undefined && t.predicate === undefined; }))), _q = _p.next(); !_q.done; _q = _p.next()) {
                                                var tag = _q.value;
                                                try {
                                                    tag.predicate = ExpressionFactory_1.ExpressionFactory.transpile(context)(tag.validate);
                                                }
                                                catch (_r) {
                                                    result.push("Unable to transpile type tag script: ".concat(JSON.stringify(tag.validate)));
                                                    tag.predicate = function () { return typescript_1.default.factory.createTrue(); };
                                                }
                                            }
                                        }
                                        catch (e_3_1) { e_3 = { error: e_3_1 }; }
                                        finally {
                                            try {
                                                if (_q && !_q.done && (_c = _p.return)) _c.call(_p);
                                            }
                                            finally { if (e_3) throw e_3.error; }
                                        }
                                    }
                                }
                                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                                finally {
                                    try {
                                        if (_o && !_o.done && (_b = _m.return)) _b.call(_m);
                                    }
                                    finally { if (e_2) throw e_2.error; }
                                }
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (_l && !_l.done && (_a = _k.return)) _a.call(_k);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                    result.push.apply(result, __spreadArray([], __read(visitor.functor(meta, explore)), false));
                    if (result.length)
                        visitor.errors.push({
                            name: meta.getName(),
                            explore: __assign({}, explore),
                            messages: __spreadArray([], __read(new Set(result)), false),
                        });
                    try {
                        for (var _s = __values(meta.aliases), _t = _s.next(); !_t.done; _t = _s.next()) {
                            var alias = _t.value;
                            validateAlias(context)(options)(visitor)(alias, explore);
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_t && !_t.done && (_d = _s.return)) _d.call(_s);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                    try {
                        for (var _u = __values(meta.arrays), _v = _u.next(); !_v.done; _v = _u.next()) {
                            var array = _v.value;
                            validateArray(context)(options)(visitor)(array.type, explore);
                        }
                    }
                    catch (e_5_1) { e_5 = { error: e_5_1 }; }
                    finally {
                        try {
                            if (_v && !_v.done && (_e = _u.return)) _e.call(_u);
                        }
                        finally { if (e_5) throw e_5.error; }
                    }
                    try {
                        for (var _w = __values(meta.tuples), _x = _w.next(); !_x.done; _x = _w.next()) {
                            var tuple = _x.value;
                            validateTuple(context)(options)(visitor)(tuple.type, explore);
                        }
                    }
                    catch (e_6_1) { e_6 = { error: e_6_1 }; }
                    finally {
                        try {
                            if (_x && !_x.done && (_f = _w.return)) _f.call(_w);
                        }
                        finally { if (e_6) throw e_6.error; }
                    }
                    try {
                        for (var _y = __values(meta.objects), _z = _y.next(); !_z.done; _z = _y.next()) {
                            var obj = _z.value;
                            validateObject(context)(options)(visitor)(obj);
                        }
                    }
                    catch (e_7_1) { e_7 = { error: e_7_1 }; }
                    finally {
                        try {
                            if (_z && !_z.done && (_g = _y.return)) _g.call(_y);
                        }
                        finally { if (e_7) throw e_7.error; }
                    }
                    try {
                        for (var _0 = __values(meta.sets), _1 = _0.next(); !_1.done; _1 = _0.next()) {
                            var set = _1.value;
                            validateMeta(context)(options)(visitor)(set, explore);
                        }
                    }
                    catch (e_8_1) { e_8 = { error: e_8_1 }; }
                    finally {
                        try {
                            if (_1 && !_1.done && (_h = _0.return)) _h.call(_0);
                        }
                        finally { if (e_8) throw e_8.error; }
                    }
                    try {
                        for (var _2 = __values(meta.maps), _3 = _2.next(); !_3.done; _3 = _2.next()) {
                            var map = _3.value;
                            validateMeta(context)(options)(visitor)(map.key, explore);
                            validateMeta(context)(options)(visitor)(map.value, explore);
                        }
                    }
                    catch (e_9_1) { e_9 = { error: e_9_1 }; }
                    finally {
                        try {
                            if (_3 && !_3.done && (_j = _2.return)) _j.call(_2);
                        }
                        finally { if (e_9) throw e_9.error; }
                    }
                    if (options.escape === true && meta.escaped !== null)
                        validateMeta(context)(options)(visitor)(meta.escaped.returns, __assign(__assign({}, explore), { escaped: true }));
                };
            };
        };
    };
    var validateAlias = function (context) {
        return function (options) {
            return function (visitor) {
                return function (alias, explore) {
                    if (visitor.aliases.has(alias))
                        return;
                    visitor.aliases.add(alias);
                    validateMeta(context)(options)(visitor)(alias.value, __assign(__assign({}, explore), { nested: alias, aliased: true }));
                };
            };
        };
    };
    var validateArray = function (context) {
        return function (options) {
            return function (visitor) {
                return function (array, explore) {
                    if (visitor.arrays.has(array))
                        return;
                    visitor.arrays.add(array);
                    validateMeta(context)(options)(visitor)(array.value, __assign(__assign({}, explore), { nested: array, top: false }));
                };
            };
        };
    };
    var validateTuple = function (context) {
        return function (options) {
            return function (visitor) {
                return function (tuple, explore) {
                    var e_10, _a;
                    if (visitor.tuples.has(tuple))
                        return;
                    visitor.tuples.add(tuple);
                    try {
                        for (var _b = __values(tuple.elements), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var elem = _c.value;
                            validateMeta(context)(options)(visitor)(elem, __assign(__assign({}, explore), { nested: tuple, top: false }));
                        }
                    }
                    catch (e_10_1) { e_10 = { error: e_10_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_10) throw e_10.error; }
                    }
                };
            };
        };
    };
    var validateObject = function (context) {
        return function (options) {
            return function (visitor) {
                return function (object) {
                    var e_11, _a;
                    if (visitor.objects.has(object))
                        return;
                    visitor.objects.add(object);
                    if (options.validate) {
                        var explore = {
                            object: object,
                            top: false,
                            property: null,
                            nested: null,
                            escaped: false,
                            aliased: false,
                        };
                        var errors = options.validate(Metadata_1.Metadata.create(__assign(__assign({}, Metadata_1.Metadata.initialize()), { objects: [object] })), explore);
                        if (errors.length)
                            visitor.errors.push({
                                name: object.name,
                                explore: explore,
                                messages: __spreadArray([], __read(new Set(errors)), false),
                            });
                    }
                    try {
                        for (var _b = __values(object.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var property = _c.value;
                            validateMeta(context)(options)(visitor)(property.value, {
                                object: object,
                                property: property.key.isSoleLiteral()
                                    ? property.key.getSoleLiteral()
                                    : {},
                                nested: null,
                                top: false,
                                aliased: false,
                                escaped: false,
                            });
                        }
                    }
                    catch (e_11_1) { e_11 = { error: e_11_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_11) throw e_11.error; }
                    }
                };
            };
        };
    };
})(MetadataFactory || (exports.MetadataFactory = MetadataFactory = {}));
